### ISSUE TITLE
Onedrive file with "～" character in name is incorrectly read as "~"

### ISSUE THREAD

**Issue by sagan** [rclone/rclone#2913]
#### What is the problem you are having with rclone?

rclone failed to list (read) filenames with "～" character in OneDrive (Business account). It incorrectly read the "～" character as "~". ( ```"～" != "~"``` )

Demo:

```
# mkdir /tmp/test && cd $_
# echo test > '～.txt'
# ls
～.txt
# rclone copy . odrive:/test/
# rclone ls odrive:/test/
        5 ~.txt
```

It does upload the file with the correct "～.txt" filename (confirmed in OneDrive web page), but read it with the wrong "~.txt" filename.

I only find this phenomenon in OneDrive (business) backend. Rclone read the correct filenames from Google Drive backend.

#### What is your rclone version (output from `rclone version`)

rclone v1.45
- os/arch: linux/amd64
- go version: go1.11.2

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

Debian sid, x64

####  Which cloud storage system are you using? (eg Google Drive)

Onedrive Business

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

see above

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

```
# rclone  -vvv ls odrive:/test
2019/01/15 07:11:13 DEBUG : rclone: Version "v1.45" starting with parameters ["rclone" "-vvv" "ls" "odrive:/test"]
2019/01/15 07:11:13 DEBUG : Using config file from "/root/.config/rclone/rclone.conf"
        5 ~.txt
2019/01/15 07:11:15 DEBUG : 5 go routines active
2019/01/15 07:11:15 DEBUG : rclone: Version "v1.45" finishing with parameters ["rclone" "-vvv" "ls" "odrive:/test"]
```

**Comment by Cnly** [comment]
This is because rclone automatically translates potentially problematic characters in file names. I had a quick look at https://support.office.com/en-us/article/invalid-file-names-and-file-types-in-onedrive-onedrive-for-business-and-sharepoint-64883a5d-228e-48f5-b3d2-eb39e07630fa#invalidfilefoldernames and it seems there are two rules related to `~`:

> Not allowed: any filename starting with ~$.
> You can’t create a folder name in SharePoint Online that begins with a tilde (~).

So it seems reasonable to make some enhancements to rclone's translation. Since there may be some compatibility issues for old files with `~` uploaded by rclone, this change should be made carefully.

@sagan Do you have any good ideas on this? :)

**Comment by sagan** [comment]
Hi,

I don't think it's the case. 

It says OneDrive disallow "```~```" in the beginning of filename in your provided link. But my problem is that rclone translate all (not only the beginning) "～" characters in filename to "~".  The "```～```" and "```~```" two are not the same characters. (Chrome highlight the both when Ctrl+F find any one in page, but they are **not** the same)

**Comment by Cnly** [comment]
@sagan I would make this clearer. Although the *current* OneDrive documentation says that not all `~` are illegal, it's possible that there *have been* more limitations on `~`, which may explain why rclone translates all `~` to `～`. And since rclone translates `~` to `～` when it's uploading files, it translates `～` back to `~` when it sees a `～` on the remote, which is why you're seeing `~` instead of the correct `～`.

**Comment by sagan** [comment]
@Cnly

Thank you for the the detailed explanation.  So how about add a command-line option which totally disable such kind of filename translations for special platforms ? (Skip the file, or throw a exception and exit, if encountering the "INVALID_FILENAME" api error when uploading a file)

There could be another way, which is "normalize" the filenames before comparing in sync / check commands. Currently, if I have a "～.txt" in OneDrive and local (or another back end, such as Google Drive) folder . They will be always considered as different files. So if I ```rclone sync /tmp/test odrive:/test```, It will re-upload the  "～.txt" to OneDrive every time.  "Normalize" the filenames will help, that's say, convert the "～" to "~" in both src and dst,  before comparing.

**Comment by sroby** [comment]
I am having the same problem right now.

It seems clear to me that the current way rclone is dealing with OneDrive-illegal characters is objectively wrong. rclone should use some sort of escape sequences instead, to identify OneDrive-illegal characters, so that the conversion properly works both ways?

**Comment by B4dM4n** [comment]
This bug should be fixed by #3148. You can try the beta build of the PR to see if it handles all cases correctly.

It will encode the normal `~` if it is the first character in a name when uploading as `～` (the FULLWIDTH variant) and undo this replacement when downloading the file from OneDrive.
A filename beginning with a `～` (FULLWIDTH) will be uploaded as `‛～` to differentiate between the the normal `~`.

#3148 currently does not distinguish between personal and business OneDrive.

**Comment by ncw** [comment]
I've merged the fix for this to master now which means it is in [the latest beta](https://beta.rclone.org) and will be released in v1.50
